src/cpp/src/visual_language/qwen2_5_vl/classes.cpp

```diff
@@ -0,0 +1,225 @@
+    const VLMConfig& vlm_config,
+    const std::filesystem::path& model_dir,
+    const std::string& device,
+    const ov::AnyMap device_config) :
+    InputsEmbedderQwen2VL(vlm_config, model_dir, device, device_config) {}
+
+InputsEmbedderQwen2_5_VL::InputsEmbedderQwen2_5_VL(
+    const VLMConfig& vlm_config,
+    const ModelsMap& models_map,
+    const Tokenizer& tokenizer, 
+    const std::filesystem::path& config_dir_path,
+    const std::string& device,
+    const ov::AnyMap device_config) :
+    InputsEmbedderQwen2VL(vlm_config, models_map, tokenizer, config_dir_path, device, device_config) {}
+
+ov::Tensor InputsEmbedderQwen2_5_VL::run_image_embeddings_merger(
+    const std::vector<EncodedImage>& images, 
+    const std::vector<size_t>& images_sequence, 
+    size_t image_id, 
+    const VLMConfig& vlm_config) {
+    
+    std::vector<ov::Tensor> image_embeds;
+    std::vector<std::array<size_t, 3>> images_grid_thw;
+    image_embeds.reserve(images.size());
+    images_grid_thw.reserve(images.size());
+    
+    for (const auto& encoded_image : images) {
+        ov::Tensor single_image_embeds = encoded_image.resized_source;
+        image_embeds.push_back(std::move(single_image_embeds));
+
+        size_t grid_t = 1;
+        size_t grid_h = encoded_image.resized_source_size.height;
+        size_t grid_w = encoded_image.resized_source_size.width;
+        images_grid_thw.push_back({grid_t, grid_h, grid_w});
+    }
+
+    std::vector<ov::Tensor> reordered_image_embeds;
+    std::vector<std::array<size_t, 3>> reordered_images_grid_thw;
+    for (size_t new_image_id : images_sequence) {
+        reordered_image_embeds.push_back(image_embeds.at(new_image_id - image_id));
+        reordered_images_grid_thw.push_back(images_grid_thw.at(new_image_id - image_id));
+    }
+
+    // Calculate cumulative sequence lengths for attention mask
+    std::vector<int32_t> cu_seqlens;
+    cu_seqlens.push_back(0);
+    int32_t cumsum = 0;
+    for (const auto& grid_thw : reordered_images_grid_thw) {
+        size_t slice_len = grid_thw.at(1) * grid_thw.at(2);
+        for (size_t t = 0; t < grid_thw.at(0); ++t) {
+            cumsum += slice_len;
+            cu_seqlens.push_back(cumsum);
+        }
+    }
+
+    // Create attention mask for vision embeddings merger model
+    size_t hidden_states_size = cumsum;
+    ov::Tensor attention_mask{ov::element::f32, {1, hidden_states_size, hidden_states_size}};
+    float* attention_mask_data = attention_mask.data<float>();
+    std::fill_n(attention_mask_data, attention_mask.get_size(), -std::numeric_limits<float>::infinity());
+
+    for (size_t i = 1; i < cu_seqlens.size(); ++i) {
+        size_t start = cu_seqlens[i-1];
+        size_t end = cu_seqlens[i];
+        for (size_t row = start; row < end; ++row) {
+            for (size_t col = start; col < end; ++col) {
+                attention_mask_data[row * hidden_states_size + col] = 0.0f;
+            }
+        }
+    }
```

Comment:
Everything below matches older qwen. Can it be factored out or did have you tried it already?

Reply:
Refactored, moved to separate methods and reused